### ISSUE TITLE
Tag handling differences between 4.x and 5.x

### ISSUE THREAD

**Issue by jstolarek** [gollum/gollum#1487]
Tags seem to be broken on 5.x branch.  Steps to reproduce:

```
mkdir -p broken_tags/foo
cd broken_tags
touch foo/bar.md
echo "[[bar]]" > Home.md
git init
git add .
git commit -m "Initial commit"
```
Result under stable version of gollum:
![gollum4](https://user-images.githubusercontent.com/1597921/76775052-b7892900-679c-11ea-84d2-7fe672310ceb.png)

Result on 5.x branch:
![gollum5](https://user-images.githubusercontent.com/1597921/76775079-bf48cd80-679c-11ea-9275-4fbbae9f878e.png)

Notice that link on the 5.x branch is broken. I looked at the 5.0 release notes wiki page but couldn't find anything that would explain why this doesn't work.

**Comment by dometto** [comment]
Thanks for the report. Could you paste the output of `gollum --versions`?

**Comment by jstolarek** [comment]
Gollum 4.1.4 and Gollum 5.0.1b (on commit f516b1cb)

**Comment by bartkamphorst** [comment]
Maybe I'm being dense, but should the tag be `[[foo/bar]]`, or `[[bar|foo/bar]]`?

**Comment by dometto** [comment]
That seems right @bartkamphorst! Although I can see where the confusion comes from, the wiki states:

```
[[Bilbo Baggins.md]]
The referenced page may exist anywhere in the git repository. Gollum will search for it and link the first such page that it finds.
```

That was likely true in some previous version of gollum, but isn't anymore: that made it utterly unpredictable which resource Gollum was going to serve. You just have to use a relative link. I'll update the wiki and the release notes.

**Comment by bartkamphorst** [comment]
Ah, that makes sense. Thanks @dometto .

**Comment by jstolarek** [comment]
Ouch. That's going to make transition to 5.x a major pain for me. So is there any difference between `[[bar|foo/bar]]` and `[bar](foo/bar)`?

**Comment by dometto** [comment]
Yes, I can see it could be a bit of a hassle converting existing wikis. But I think in the end it helps keep one's information organized, and it makes it a whole lot easier to keep supporting gollum if we can write tests being certain of which resources we are specifying.

However, I just noticed that we support relative links for images in the same directory as the linking file. So if you have `greek/Bilbo` and `greep/alpha.jpg`, in Bilbo.md you can link simply using `[[alpha.jpg]]`. This isn't working for page links, though: ``[[Foo.md]]` is being interpreted as an absolute path. I'll see if we can do something about that, which would hopefully make things a bit easier.

**Comment by dometto** [comment]
I implemented support for relative page links in https://github.com/gollum/gollum-lib/pull/360. Hopefully that will make things a little easier.

**Comment by dometto** [comment]
Documented the (new) behavior in https://github.com/gollum/gollum/wiki/5.0-release-notes#no-global-file-names

**Comment by bartkamphorst** [comment]
> So is there any difference between `[[bar|foo/bar]]` and `[bar](foo/bar)`?

Interestingly, yes! The latter still produces a link without file extension (e.g., `http://localhost:4567/foo/bar`. @dometto Bug, right? Should be `http://localhost:4567/foo/bar.md` I think?

**Comment by jstolarek** [comment]
>  Should be http://localhost:4567/foo/bar.md I think?

I truly hope not, because that would mean breaking even more of backward compatibility making a transition even more painful.

Also, to comment on what @dometto said:

>  But I think in the end it helps keep one's information organized

I have a moderately-sized personal wiki, close to 200 subpages, so not extremely large but also not that small. I always made sure that my subpages have different names and have relied heavily on the tags feature working the way it did so far. The transition to a newer format will be a major pain - if it happens, because now I realize that converting all the links will take many hours of work which I'm not sure I am willing to invest.

**Comment by dometto** [comment]
> Interestingly, yes! The latter still produces a link without file extension (e.g., http://localhost:4567/foo/bar. @dometto Bug, right? Should be http://localhost:4567/foo/bar.md I think?

Not a bug, I think, as the `[]()` style links are handled by the markdown/markup renderer -- which is just producing links to whatever the user literally puts in.

@jstolarek thanks for the feedback. Would it address your concerns if we included a flag for maintaining the old behavior?

**Comment by jstolarek** [comment]
@dometto - that would definitely be helpful. I also wonder about the rationale for changing the behaviour of tags. Have the users complained about how this feature works?

**Comment by dometto** [comment]
> @dometto - that would definitely be helpful. I also wonder about the rationale for changing the behaviour of tags. Have the users complained about how this feature works?

Yes, as it makes for terrible performance in larger wikis: for each link, we potentially need to walking over (almost) all files in the repository and see if it matches. We'll make sure to include that explanation in the release notes.

>  Would it address your concerns if we included a flag for maintaining the old behavior?

Alternatively, we're looking into packaging along a conversion script that will automatically replace broken links -- although that would of course work only for `[[gollum]]` tags, and not for `[markdown](link)` tags.

**Comment by jstolarek** [comment]
> Alternatively, we're looking into packaging along a conversion script that will automatically replace broken links -- although that would of course work only for [[gollum]] tags, and not for [markdown](link) tags.

That would be perfect!

**Comment by jstolarek** [comment]
Then again I'm not sure whether it makes sense for you to put effort into a migration script. I have a specific requirements of how it should work, but I doubt these are universal requirements that would work for other users.

**Comment by bartkamphorst** [comment]
> the []() style links are handled by the markdown/markup renderer

Thanks for clarifying. 👍

**Comment by dometto** [comment]
@jstolarek there is now a script that migrates 4.x-style 'global' page link tags to 5.x-style absolute paths, where required:

https://gist.github.com/dometto/6ed2897e35e8c2832802178bcc491686

Would you mind giving it a spin?

The script needs to live in the directory of your gollum 5.x install for now. Make sure to run `bundle update` before trying it.

To run, we recommend:

1. Cloning your wiki's repository
1. First do a dry run on your cloned repo: `bundle exec ruby gollum-migrator.rb /path/to/cloned/repo --dry-run`
1. If all looks sane, run it for real: `bundle exec ruby gollum-migrator.rb /path/to/cloned/repo`
1. Do a `git diff` to inspect the changes (the script doesn't automatically commit the changes).
1. Commit the changes if all looks sane, and push/pull them back into your original repo.

**Comment by dometto** [comment]
To add to that: the script also supports the gollum `--page-file-dir` and `--config` options.

**Comment by dometto** [comment]
Updated the script to also fix disparities caused by the fact that 5.x checks paths case-sensitively.

**Comment by jstolarek** [comment]
So the script does work, but it's not suitable for my needs. The major reason why I resorted to using tags was that having relative links to anywhere in the wiki made it very easy for me re-organise the structure of the pages whenever I wanted. Having links to absolute page locations is a complete opposite of that and not an option for me.

As a bit of background, I try to have my wiki organized in a tree-like structure as much as possible. This means that an overwhelming majority of links in my wiki either points to a page in a subdirectory or, conversely, to a page in a direct super-directory. This makes it acceptably easy to reorganise the structure of pages, e.g. moving or renaming a subdirectory with a bunch of pages inside usually means changing only a couple of links, since the links on pages within a directory remain unaffected - they are relative. That's why I don't want to use absolute paths.

**Comment by dometto** [comment]
@jstolarek thanks for the further explanation. We've decided on the following:

1. We're adding a command line option to maintain the old behavior. https://github.com/gollum/gollum-lib/pull/363/files
  * we've decided against making this the default, as a) it is costly performance wise and b) other wikis (DokuWiki and Wiki.JS, for example) also don't have this behavior.
2. We'll package the migration script to encourage users who can migrate to do so
3. The script will have an option allowing to users to convert links to *relative* links where possible.

**Comment by jstolarek** [comment]
Thanks. This sounds like a reasonable migration plan. One thing that would also be useful to me is being able to check whether there are any broken internal links on a wiki. I think I can do a huge chunk of my migration (about 80%) with smart usage of `sed`. But then it would be useful to have a way of checking whether all the links work.

**Comment by dometto** [comment]
The migration script already provides that functionality. It prints an `[INFO]` line when encountering a broken link. If you run the script with `--dry-run`, it won't actually touch any of your files, but print the info anyway.

**Comment by jstolarek** [comment]
I have managed to migrate all of my wiki to the new link format. I found a solution that made it a lot simpler. Back in the day I used to use normal `[]()` links and then did a massive switch to tags when I discovered how they (used to) work. What I did now was simply revert the commits. That did most of the job. I was left with fixing a handful of links but thanks to the migration script locating what needs to be fixed was very simple.
